x-pack/plugins/infra/public/components/logging/log_text_stream/loading_item_view.tsx

```diff
@@ -6,144 +6,272 @@
-          </ProgressMessage>
-          {onLoadMore ? (
-            <EuiButtonEmpty size="xs" onClick={onLoadMore} iconType="refresh">
-              <FormattedMessage
-                id="xpack.infra.logs.loadAgainButtonLabel"
-                defaultMessage="Load again"
-              />
-            </EuiButtonEmpty>
-          ) : null}
-        </ProgressEntry>
-      );
-    } else {
-      return null;
-    }
+    const shouldShowCta = !hasMore && !isStreaming;
+
+    const extra = (
+      <EuiFlexGroup justifyContent="center" alignItems="center" gutterSize="m">
+        {isLoading || isStreaming ? (
+          <ProgressSpinner kind={isStreaming ? 'streaming' : 'loading'} />
+        ) : shouldShowCta ? (
+          <ProgressCta
+            position={position}
+            onStreamStart={onStreamStart}
+            onExtendRange={onExtendRange}
+            startDateExpression={startDateExpression}
+            endDateExpression={endDateExpression}
+          />
+        ) : null}
+      </EuiFlexGroup>
+    );
+
+    return (
+      <ProgressEntryWrapper className={className} position={position}>
+        {position === 'start' ? <>{extra}</> : null}
+        <ProgressMessage timestamp={timestamp} relative={position === 'end' && isStreaming} />
+        {position === 'end' ? <>{extra}</> : null}
+      </ProgressEntryWrapper>
+    );
   }
 }
 
-interface ProgressEntryProps {
-  alignment: 'top' | 'bottom';
-  className?: string;
-  color: 'subdued' | 'primary';
-  isLoading: boolean;
-}
+const ProgressEntryWrapper = euiStyled.div<{ position: Position }>`
+  padding-left: ${props => props.theme.eui.euiSizeS};
+  padding-top: ${props =>
+    props.position === 'start' ? props.theme.eui.euiSizeL : props.theme.eui.euiSizeM};
+  padding-bottom: ${props =>
+    props.position === 'end' ? props.theme.eui.euiSizeL : props.theme.eui.euiSizeM};
+`;
 
-const ProgressEntry: React.FC<ProgressEntryProps> = props => {
-  const { alignment, children, className, color, isLoading } = props;
+const ProgressMessage: React.FC<{ timestamp?: number; relative?: boolean }> = ({
+  timestamp,
+  relative = false,
+}) => {
+  let message;
 
-  // NOTE: styled-components seems to make all props in EuiProgress required, so this
-  // style attribute hacking replaces styled-components here for now until that can be fixed
-  // see: https://github.com/elastic/eui/issues/1655
-  const alignmentStyle =
-    alignment === 'top' ? { top: 0, bottom: 'initial' } : { top: 'initial', bottom: 0 };
+  if (timestamp) {
+    const formattedTimestamp = relative ? (
+      <FormattedRelative units="second" value={timestamp} updateInterval={1} />
+    ) : (
+      <FormattedTime value={timestamp} {...TIMESTAMP_FORMAT} />
+    );
+    message = (
+      <FormattedMessage
+        id="xpack.infra.logs.showingEntriesUntilTimestamp"
+        defaultMessage="Showing entries until {timestamp}"
+        values={{ timestamp: formattedTimestamp }}
+      />
+    );
+  } else {
```

Comment:
Can you help me understand when `timestamp` would be falsy? When would we display this message?

Reply:
mmm, good point. I think I'm being unnecessarily defensive